### ISSUE TITLE
Code clarification on prediction in model.py

### ISSUE THREAD

**Issue by zyzhang1130** [karpathy/nanoGPT#200]
If I understand it correctly, at line 193 of model.py, what it does is to predict the probabilities of the subsequent tokens and compare them against the ground truth (line 194). What I don't understand is, it seems the logits of the entire block length were obtained in one go by passing through lm_head which is just a linear layer. So it seems that the next token predictions for the tokens within the block length were already implicitly done, since in order to get the next token, the prediction needs to be conditioned on the previous predicted token. I don't understand how it can be done with just a linear layer, and what should I do if I want all the intermediate logits conditioned on all possible preceding sequences.

**Comment by Andrei-Aksionov** [comment]
Hello @zyzhang1130 

It's done with help of [masking](https://github.com/karpathy/nanoGPT/blob/master/model.py#L76) in the attention layer.
So the token at the 1st position has information about itself, token at the second - about the 1st and itself and so on.
And `lm_head` is applied on each token embedding individually, so `lm_head` processes each token with encoded information of all tokens to the left of it and of course itself. 
That's why it's possible to do it with just a linear layer: all the information about the token and it's context is already calculated before applying `lm_head`.

> ... the prediction needs to be conditioned on the previous predicted token

Not quite. We use previously predicted tokens during token generation step. During training we are trying to predict new token given it's context to the left of it.

Sorry if I misunderstood your question.

**Comment by Andrei-Aksionov** [comment]
Let's say:
- text = "abcd"
- x = [a, b, c]
- y = [b, c, d]
(y - is the same as x with shift by 1 token to the right)

so given x=[a] we want the model to predict [b],
given x=[a, b] -> [c]
given x=[a, b, c] -> [d]

and logits will contain prediction for each token given it's context:
logits = [b, c, d]
(it's for simplicity, in reality of course logits contain values for each element in the vocabulary for each token: block_size * vocab_size)

and now we can compare it with `y` to compute the loss

**Comment by zyzhang1130** [comment]
I see. So the next token is always conditioned on the ground truth preceding the tokens. Thanks a lot for addressing my confusion.

**Comment by zyzhang1130** [comment]
Hi, just one follow-up question: in line 197 of `model.py`, 'only forward the lm_head on the very last position'. Why it does not affect the conditioning when the input is only the last token (I assume the output logits here still need to be conditioned on all preceding tokens)?

**Comment by zyzhang1130** [comment]
Hi, just one follow-up question: in line 197 of model.py, 'only forward the lm_head on the very last position'. Why it does not affect the conditioning when the input is only the last token (I assume the output logits here still need to be conditioned on all preceding tokens)?

**Comment by zyzhang1130** [comment]
Hi, just one follow-up question: in line 197 of model.py, 'only forward the lm_head on the very last position'. Why it does not affect the conditioning when the input is only the last token (I assume the output logits here still need to be conditioned on all preceding tokens)?

**Comment by Andrei-Aksionov** [comment]
Hello @zyzhang1130 

If targets is not None, then it's either training or evaluation step, if is None - new token generation/sampling.
When we do training/evaluation we want to measure loss value and for it we need to see predictions for each token in the block (like I stated in the previous answer).

When doing sampling and let's say we already have sampled tokens [a,b,c], then we don't care what the model would outputted having [a] or [a,b], we care only what the model would outputted after [a,b,c]. That's why we apply language model head on the very last token in the block. Token at the last position in the block will contain all the information about it's context via attention layers.

For clarification: language model head only maps token's embedding onto vocabulary, it uses information about the token and all it's preceding tokens (that was generated before), but not generates it. `lm_head` is applied on each token embedding individually, so it "knows" about preceding tokens only via the information that is provided to it in the token embedding itself.

**Comment by zyzhang1130** [comment]
> `lm_head` is applied on each token embedding individually, so it "knows" about preceding tokens only via the information that is provided to it in the token embedding itself.

Thanks again for the detailed explanation. For the quoted statement, I'm just curious how it is achieved in `model.py`?

**Comment by Andrei-Aksionov** [comment]
It happens [here](https://github.com/karpathy/nanoGPT/blob/master/model.py#L75).
Or to be more specific during calculation of attention scores. Here is one more [link](https://github.com/Andrei-Aksionov/nanoGPTplus/blob/main/src/model/gpt_language_model/attention.py#L94) to my repo (it contains some additional comments of what's going on).

If to take a look at the scheme you can see that there are arrows that start from the bottom and go straight up. That's trying to show that through all the steps tokens are independent from each other and each doesn't one interact with others*.
![image](https://user-images.githubusercontent.com/58434077/224548777-68a780bb-81e0-4f1a-aca1-fbcb03d092e0.png)
*Except one step: in attention layer when we calculate attention scores through multiplying queries and keys. Then we use those scores to make weighted averaging of values.

So calculation of information about other tokens can happens only in this place 😄 

If you are still confused about this whole weighted averaging action, let's look at the example.
Sentence: I have a piano and it's keys are broken.
Now let's focus on token "keys".
When calculating attention scores the biggest values would be for tokens "keys", "piano",  maybe values like 0.6, 0.35, and 0.05 will be spread out among others. That means that in order to represent token "keys", we need to take 60% of it's value, 35% of token "piano" and 5% of others. That's why we multiply attention scores with values.

Now, why it's important. Because work "key" might mean different things depending on it's context: it might be a piano key, computer keyboard key, key that opens a door, .... So after attention token's embedding will contain this information.
Plus it's processed with Feed-Forward layer (also applied on each token individually), and it's all done multiple time if we have multiple layers. Plus information from positional encoding, so the token also stores info of where it is in the sequence. And yeah, don't forget about multihead attention, where each head learns something different.

So if to come back to your original question related to how a simple linear layer (`lm_head`) knows about all the preceding tokens. The answer that it doesn't know and it doesn't have to: all the info is already provided.

---
Still have questions?
Look at this [article](https://jalammar.github.io/illustrated-transformer/)(it has colorful pictures 😃).
In my opinion the most intuitive exaplantion of key, query and value is done in [google's tutorial](https://www.tensorflow.org/text/tutorials/transformer#attention_refresher).

**Comment by zyzhang1130** [comment]
Awesome! Thank you so much for the answer and the resources provided!
